rust2/core/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+//!
+//! # Native Rust drivers
+//!
+//! Native Rust drivers will implement the abstract API consisting of the traits:
+//! - [Driver]
+//! - [Database]
+//! - [Connection]
+//! - [Statement]
+//!
+//! For drivers implemented in Rust, using these will be more efficient and
+//! safe, since it avoids the overhead of going through C FFI.
+//!
+//! # Driver Manager
+//!
+//! The [driver_manager] module allows loading drivers exposing the C API,
+//! either from an initialization function (link-time, either static or dynamic)
+//! or by dynamically finding such a function in a dynamic library (run-time).
+
+// TODO(alexandreyc): uncomment these lines during follow-up PRs
+// pub mod driver_exporter;
+// pub mod driver_manager;
+// pub use ffi::FFI_AdbcDriverInitFunc as AdbcDriverInitFunc;
+
+pub mod error;
+pub mod ffi;
+pub mod options;
+pub mod schemas;
+
+use arrow::datatypes::Schema;
+use arrow::record_batch::{RecordBatch, RecordBatchReader};
+
+use error::Result;
+use options::{OptionConnection, OptionDatabase, OptionStatement, OptionValue};
+
+/// Ability to configure an object by setting/getting options.
+pub trait Optionable {
+    type Option: AsRef<str>;
+
+    /// Set a post-init option.
+    fn set_option(&mut self, key: Self::Option, value: OptionValue) -> Result<()>;
+
+    /// Get a string option value by key.
+    fn get_option_string(&self, key: Self::Option) -> Result<String>;
+
+    /// Get a bytes option value by key.
+    fn get_option_bytes(&self, key: Self::Option) -> Result<Vec<u8>>;
+
+    /// Get an integer option value by key.
+    fn get_option_int(&self, key: Self::Option) -> Result<i64>;
+
+    /// Get a float option value by key.
+    fn get_option_double(&self, key: Self::Option) -> Result<f64>;
+}
+
+/// A handle to an ADBC driver.
+pub trait Driver {
+    type DatabaseType: Database;
+
+    /// Allocate and initialize a new database without pre-init options.
+    fn new_database(&self) -> Result<Self::DatabaseType>;
+
+    /// Allocate and initialize a new database with pre-init options.
+    fn new_database_with_opts(
+        &self,
+        opts: impl Iterator<Item = (OptionDatabase, OptionValue)>,
+    ) -> Result<Self::DatabaseType>;
+}
+
+/// A handle to an ADBC database.
+///
+/// Databases hold state shared by multiple connections. This typically means
+/// configuration and caches. For in-memory databases, it provides a place to
+/// hold ownership of the in-memory database.
+///
+/// Databases must be kept alive as long as any connections exist.
+pub trait Database: Optionable<Option = OptionDatabase> {
+    type ConnectionType: Connection;
+
+    /// Allocate and initialize a new connection without pre-init options.
+    fn new_connection(&self) -> Result<Self::ConnectionType>;
+
+    /// Allocate and initialize a new connection with pre-init options.
+    fn new_connection_with_opts(
+        &self,
+        opts: impl Iterator<Item = (options::OptionConnection, OptionValue)>,
```

Comment:
> We could also define a non-fieldless enum and accept an iterator, but that allows duplicate values for the same key, so I really think a product type is better here.

That's what I did in a first version but it wasn't a great fit because we also need to query option values with `get_option_*` and in that case we only need the option key and not the value (that would force us to provide a sentinel in the non-fieldless enum variant).

Regarding the product type, IMO it's not a great fit either:
1. Databases not necessarily need all options. For instance, SQLite only needs `uri`, it even does not recognize `username` nor `password` (we could wrap all the fields in `Option` though).
2. Everytime a new "standard" key is added we'll have to add a new field to this struct.
3. It's harder for the user to split between pre-init and post-init options, because in that case he would need to define two instances of the struct with non-overlapping non-`None` fields.

The only downside I see with the current solution is the possibility to provide duplicated keys... We could document that "last write win" strategy.

Reply:
I'm not sure to understand what you mean. Can you elaborate?

Reply:
I've tried to come up with a similar design that provides all the required capabilities but I'm feeling stuck...

A few points:
- `new_connection_with_opts` takes *connection* options, not database options (same logic applies to other `new_` methods). This would forces us to duplicate the associated type into two traits.
- How do we get/set non-standard option in a generic way? That's required for the driver exporter. We also need to be able to set standard option in a generic way (but this can be done by adding `set_uri` method).
- According to the spec, we must be able to set an Int option and get it back as String (or any other combination of types)
- We must assume that each standard option key is associated with a specific type for its value. It seems to the case as of now but it's not explicit in the spec.

I will put more thougth into it, but I've tried different designs before settling on this one. While not perfect, it's relatively simple and can accomodate the C API...

Reply:
> we must be able to set an Int option and get it back as String (or any other combination of types)

this isn't strictly required, fwiw

Reply:
> We must assume that each standard option key is associated with a specific type for its value. It seems to the case as of now but it's not explicit in the spec.

I don't think this will necessarily hold, either. 


Also, re: the original proposal, not all drivers necessarily have or require uri, username, or password fields; the latter two would never apply to SQLite, for instance, and SQLite doesn't require a URI

Reply:
> I will put more thougth into it, but I've tried different designs before settling on this one. While not perfect, it's relatively simple and can accomodate the C API...

Sounds good, it's probably easier to discuss API designs here once we have some real implementations.

> Also, re: the original proposal, not all drivers necessarily have or require uri, username, or password fields; the latter two would never apply to SQLite, for instance, and SQLite doesn't require a URI

If you are referring to what I suggested, then yes, the field types should all be wrapped in `Option` as suggested by @alexandreyc:

> (we could wrap all the fields in Option though).

Reply:
So, I propose to go forward with this design for now and reconsider the question when we have more substance to discuss. Are you okay?